File: theme identification/python/vector_db.py
import chromadb
import os
import json
import uuid
import numpy as np
from typing import List, Dict, Any, Optional
from sentence_transformers import SentenceTransformer

# Initialize the embedding model
model = SentenceTransformer('all-MiniLM-L6-v2')

# Initialize ChromaDB client
client = chromadb.Client()

# Create a collection for our documents
collection = client.create_collection(name="documents")

def embed_text(text: str) -> List[float]:
    """Generate embeddings for a text string"""
    return model.encode(text).tolist()

def add_document(doc_id: str, content: str, metadata: Dict[str, Any]) -> str:
    """Add a document to the vector database"""
    # Generate embeddings for the document
    embeddings = embed_text(content)
    
    # Add document to the collection
    collection.add(
        ids=[doc_id],
        embeddings=[embeddings],
        metadatas=[metadata],
        documents=[content]
    )
    
    return doc_id

def search_documents(query: str, n_results: int = 5) -> List[Dict[str, Any]]:
    """Search for documents similar to the query"""
    # Generate embeddings for the query
    query_embedding = embed_text(query)
    
    # Search the collection
    results = collection.query(
        query_embeddings=[query_embedding],
        n_results=n_results
    )
    
    # Format results
    formatted_results = []
    for i in range(len(results['ids'][0])):
        formatted_results.append({
            'id': results['ids'][0][i],
            'document': results['documents'][0][i],
            'metadata': results['metadatas'][0][i],
            'distance': results['distances'][0][i] if 'distances' in results else None
        })
    
    return formatted_results

def chunk_document(content: str, metadata: Dict[str, Any], chunk_size: int = 1000, overlap: int = 200) -> List[str]:
    """Split a document into chunks for better semantic search"""
    chunks = []
    chunk_ids = []
    
    # Simple chunking by character count
    for i in range(0, len(content), chunk_size - overlap):
        chunk = content[i:i + chunk_size]
        if len(chunk) < 100:  # Skip very small chunks
            continue
            
        chunk_id = f"{metadata['id']}_chunk_{len(chunks)}"
        chunk_metadata = {**metadata, 'chunk_id': chunk_id, 'chunk_index': len(chunks)}
        
        # Add the chunk to the vector database
        add_document(chunk_id, chunk, chunk_metadata)
        
        chunks.append(chunk)
        chunk_ids.append(chunk_id)
    
    return chunk_ids

def process_document(content: str, filename: str, doc_id: Optional[str] = None) -> str:
    """Process a document and add it to the vector database"""
    if doc_id is None:
        doc_id = str(uuid.uuid4())
        
    metadata = {
        'id': doc_id,
        'filename': filename,
        'length': len(content)
    }
    
    # For longer documents, chunk them
    if len(content) > 1000:
        chunk_ids = chunk_document(content, metadata)
        metadata['chunk_ids'] = chunk_ids
    else:
        add_document(doc_id, content, metadata)
    
    return doc_id

def get_document_by_id(doc_id: str) -> Dict[str, Any]:
    """Retrieve a document by its ID"""
    results = collection.get(ids=[doc_id])
    
    if not results['ids']:
        return None
        
    return {
        'id': results['ids'][0],
        'document': results['documents'][0],
        'metadata': results['metadatas'][0]
    }

# Command-line interface for testing
if __name__ == "__main__":
    import sys
    
    command = sys.argv[1] if len(sys.argv) > 1 else None
    
    if command == "add":
        # Example: python vector_db.py add "document content" "filename.txt"
        content = sys.argv[2]
        filename = sys.argv[3]
        doc_id = process_document(content, filename)
        print(json.dumps({"success": True, "doc_id": doc_id}))
        
    elif command == "search":
        # Example: python vector_db.py search "query string" 5
        query = sys.argv[2]
        n_results = int(sys.argv[3]) if len(sys.argv) > 3 else 5
        results = search_documents(query, n_results)
        print(json.dumps({"success": True, "results": results}))
        
    elif command == "get":
        # Example: python vector_db.py get "doc_id"
        doc_id = sys.argv[2]
        document = get_document_by_id(doc_id)
        print(json.dumps({"success": True, "document": document}))
        
    else:
        print(json.dumps({"success": False, "error": "Invalid command"}))

File: theme identification/lib/vector-db.ts
import type { Document } from "@/types/document"

// Simple vector database implementation for browser environments
export class VectorDatabase {
  private documents: Map<string, DocumentVector> = new Map()
  private embeddings: Map<string, number[]> = new Map()
  private model: any = null
  private modelLoading = false
  private modelLoaded = false

  // Initialize the vector database
  async initialize(): Promise<boolean> {
    try {
      // We'll use a simple mock embedding function for now
      // In a real implementation, you would load a TensorFlow.js model here
      this.modelLoaded = true
      return true
    } catch (error) {
      console.error("Failed to initialize vector database:", error)
      return false
    }
  }

  // Check if the database is ready
  isReady(): boolean {
    return this.modelLoaded
  }

  // Add a document to the vector database
  async addDocument(document: Document, content: string): Promise<boolean> {
    try {
      if (!this.modelLoaded) {
        await this.initialize()
      }

      // Generate a simple embedding for the document
      const embedding = await this.generateEmbedding(content)

      // Store the document and its embedding
      const docVector: DocumentVector = {
        id: document.id,
        content,
        metadata: {
          filename: document.filename,
          fileType: document.fileType,
          pages: document.pages,
        },
        chunks: [],
      }

      // For longer documents, split into chunks
      if (content.length > 500) {
        const chunks = this.chunkDocument(content, document.id)
        docVector.chunks = chunks

        // Store embeddings for each chunk
        for (const chunk of chunks) {
          const chunkEmbedding = await this.generateEmbedding(chunk.content)
          this.embeddings.set(chunk.id, chunkEmbedding)
        }
      }

      // Store the document and its main embedding
      this.documents.set(document.id, docVector)
      this.embeddings.set(document.id, embedding)

      return true
    } catch (error) {
      console.error("Error adding document to vector database:", error)
      return false
    }
  }

  // Split a document into chunks
  private chunkDocument(content: string, documentId: string): DocumentChunk[] {
    const chunks: DocumentChunk[] = []
    const chunkSize = 500
    const overlap = 100

    for (let i = 0; i < content.length; i += chunkSize - overlap) {
      const chunkContent = content.substring(i, i + chunkSize)
      if (chunkContent.length < 100) continue // Skip very small chunks

      chunks.push({
        id: `${documentId}_chunk_${chunks.length}`,
        content: chunkContent,
        metadata: {
          parentId: documentId,
          chunkIndex: chunks.length,
        },
      })
    }

    return chunks
  }

  // Generate an embedding for text
  private async generateEmbedding(text: string): Promise<number[]> {
    // This is a very simple mock embedding function
    // In a real implementation, you would use a proper embedding model

    // Create a simple hash-based embedding (not suitable for real semantic search)
    const embedding: number[] = new Array(128).fill(0)

    // Generate some values based on the text content
    for (let i = 0; i < text.length; i++) {
      const charCode = text.charCodeAt(i)
      embedding[i % embedding.length] += charCode / 100
    }

    // Normalize the embedding
    const magnitude = Math.sqrt(embedding.reduce((sum, val) => sum + val * val, 0))
    return embedding.map((val) => val / magnitude)
  }

  // Search for documents similar to a query
  async searchDocuments(query: string, numResults = 5): Promise<SearchResult[]> {
    try {
      if (!this.modelLoaded) {
        await this.initialize()
      }

      // Generate embedding for the query
      const queryEmbedding = await this.generateEmbedding(query)

      // Calculate similarity scores for all documents and chunks
      const scores: { id: string; score: number; isChunk: boolean }[] = []

      // Score all documents and chunks
      for (const [id, embedding] of this.embeddings.entries()) {
        const similarity = this.cosineSimilarity(queryEmbedding, embedding)
        const isChunk = id.includes("_chunk_")
        scores.push({ id, score: similarity, isChunk })
      }

      // Sort by similarity score (descending)
      scores.sort((a, b) => b.score - a.score)

      // Get the top results
      const results: SearchResult[] = []
      const seenParentIds = new Set<string>()

      for (const { id, score, isChunk } of scores) {
        if (results.length >= numResults) break

        if (isChunk) {
          // This is a document chunk
          const chunkId = id
          const parentId = id.split("_chunk_")[0]

          // Skip if we already have a result from this parent document
          if (seenParentIds.has(parentId)) continue

          const document = this.documents.get(parentId)
          if (!document) continue

          // Find the chunk
          const chunk = document.chunks.find((c) => c.id === chunkId)
          if (!chunk) continue

          results.push({
            id: chunkId,
            documentId: parentId,
            content: chunk.content,
            metadata: {
              ...document.metadata,
              chunkIndex: chunk.metadata.chunkIndex,
            },
            score,
          })

          seenParentIds.add(parentId)
        } else {
          // This is a full document
          const document = this.documents.get(id)
          if (!document || seenParentIds.has(id)) continue

          results.push({
            id,
            documentId: id,
            content: document.content,
            metadata: document.metadata,
            score,
          })

          seenParentIds.add(id)
        }
      }

      return results
    } catch (error) {
      console.error("Error searching vector database:", error)
      return []
    }
  }

  // Calculate cosine similarity between two vectors
  private cosineSimilarity(a: number[], b: number[]): number {
    if (a.length !== b.length) return 0

    let dotProduct = 0
    let magnitudeA = 0
    let magnitudeB = 0

    for (let i = 0; i < a.length; i++) {
      dotProduct += a[i] * b[i]
      magnitudeA += a[i] * a[i]
      magnitudeB += b[i] * b[i]
    }

    magnitudeA = Math.sqrt(magnitudeA)
    magnitudeB = Math.sqrt(magnitudeB)

    if (magnitudeA === 0 || magnitudeB === 0) return 0

    return dotProduct / (magnitudeA * magnitudeB)
  }
}

// Singleton instance
let vectorDB: VectorDatabase | null = null

// Get the vector database instance
export function getVectorDB(): VectorDatabase {
  if (!vectorDB) {
    vectorDB = new VectorDatabase()
    vectorDB.initialize()
  }
  return vectorDB
}

// Types
interface DocumentVector {
  id: string
  content: string
  metadata: {
    filename: string
    fileType: string
    pages: number
  }
  chunks: DocumentChunk[]
}

interface DocumentChunk {
  id: string
  content: string
  metadata: {
    parentId: string
    chunkIndex: number
  }
}

export interface SearchResult {
  id: string
  documentId: string
  content: string
  metadata: any
  score: number
}
